gurklang/builtin_values.py

```diff
@@ -103,6 +103,89 @@ def close(stack: T[V, T[V, S]], scope: Scope, fail: Fail):
+        captures: list[tuple[int, Value]] = []
+        variables: dict[str, Value] = {}
+        for nested_pattern, nested_value in zip(pattern.values, value.values):
+            matches, new_captures, new_variables = _matches_impl(nested_pattern, nested_value, fail)
+            if not matches:
+                return False, [], {}
+            if variables.keys() & new_variables.keys():
+                fail(f'duplicate variable name in pattern: {variables.keys() & new_variables.keys()!r}')
+            captures.extend(new_captures)
+            variables.update(new_variables)
+        return True, captures, variables
+    elif isinstance(pattern, Atom):
+        label = pattern.value
+        if label.startswith(':') and isinstance(value, Atom) and value.value == label[1:]:
+            return True, [], {}
+        elif frozenset(label) == {'.'}:
+            return True, [(len(label), value)], {}
+        else:
+            return True, [], {label: Code([Put(value)], closure=None)}
+    elif pattern == value:
+        return True, [], {}
+    return False, [], {}
+
+
+def _matches(pattern: Vec, stack: Stack, fail: Fail) -> tuple[bool, Stack, dict[str, Value]]:
+    captures: list[tuple[int, Value]] = []
+    variables: dict[str, Value] = {}
+    original_stack = stack
+    for inner_pattern in reversed(pattern.values):
+        top, stack = stack
+        matches, stack_slots, new_vars = _matches_impl(inner_pattern, top, fail)
+        if not matches:
+            return False, original_stack, {}
+        if new_vars.keys() & variables.keys():
+            fail(f'duplicate variable name in pattern: {variables.keys() & new_vars.keys()!r}')
+        captures.extend(stack_slots)
+        variables.update(new_vars)
+    print(captures)
+    captures.sort(key=itemgetter(0), reverse=True)
+    print(captures)
+    return (
+        True,
+        _stack_extend(stack, (el for _, el in reversed(captures))),
+        variables
+    )
+
+
+def _stack_extend(stack: Stack, elems: Iterable[Value]) -> Stack:
+    for elem in elems:
+        stack = (elem, stack)
+    return stack
+
+
+@module.register()
+def case(stack: T[V, S], scope: Scope, fail: Fail):
+    sentinel = Atom(object())  # type: ignore
+    fun, rest = stack
+    new_stack, _ = vm.call((sentinel, rest), scope, fun)
+    cases = []
+    while not (isinstance(new_stack[0], Atom) and new_stack[0].value is sentinel.value):
+        next_elem, new_stack = new_stack
+        cases.append(next_elem)
+    if len(cases) % 2 == 1:
+        fail('odd number of forms in case expression, there must be exactly one function per pattern')
+    for action, pattern in zip(cases[::2], cases[1::2]):
+        if not isinstance(pattern, Vec):
+            fail(f'a pattern must be a vector, not {pattern!r}')
+        matched, new_stack, new_variables = _matches(pattern, rest, fail)
+        if matched:
+            return vm.call(new_stack, scope.with_members(new_variables), action)
```

Comment:
`vm.call` shouldn't be called from inside native functions, because this kills the stacklessness and will lead to recursion errors. 

We'll have to do something like this:
```py
@module.register()
def __get_case(stack: T[V, S], scope: Scope, fail: Fail):
    ...
    def fn(stack, scope):
        return (action, new_stack), scope.with_members(new_variables)
    return NativeFunction(fn)
    
# { get-case ! } parent-scope :case jar
module.add(
    "case",
    Code([CallByName("--get-case"), CallByValue()], name="case", flags=CodeFlags.PARENT_SCOPE)
)
```